### ISSUE TITLE
Allow injecting to private fields?

### ISSUE THREAD

**Issue by kevinresol** [jasononeil/dodrugs#2]
Is there any reason to force exposing the variables just for injection?

**Comment by jasononeil** [comment]
Good question! I always ask the same thing when using minject, it seems like an annoying limitation.

When I started creating this I think I finally understood the reason: so that you can always perform injection manually.  If we allow injecting on a private member, then suddenly dependency injection becomes something that is only possible with this specific library, and that class cannot be created manually (without using the dependency injector).

The way to keep the variables private is to use constructor injection, but of course that is quite verbose:

``` haxe
class Person {
    var name:String;
    var age:Int;
    @inject public function new( name:String, age:Int ) {
        this.name = name;
        this.age = age;
    }
}
```

What do you think of this reasoning?  If you make a good argument I can change my mind :)

**Comment by kevinresol** [comment]
Isn't that a choice rather than a requirement? If a library writer wants his class to be inited without DI he can choose to publicize the vars. But in my case I would like to have my class inited _only through_ DI, so I would like to use private vars but there is no choice for that. The constructor method works, but it tripled the amount of typing.

**Comment by jasononeil** [comment]
How would you feel about allowing it with a compilation define: `-D dodrugs_in_private` or something?

I think I'd like to allow it, but discourage it's use in public libraries. So use it in your own codebase, but discourage using it in a haxelib.

What do you think?

Otherwise we can allow it but just say in the readme when it is and isn't a good idea to use it.

**Comment by kevinresol** [comment]
I am not very sure either. Because my initial thought is kind of totally opposite to yours. I think that when I design a class that is supposed to get dependencies injected, then they should just be injected and should not be manually manipulated in any ways. Maybe I am wrong.

cc @back2dos: you usually have good ideas on topics like this.

**Comment by back2dos** [comment]
If you're worried about verbosity, then tink_lang will solve it for you:

``` haxe
@:tink class Person {
    var name:String = _;
    var age:Int = _;
    @inject function new() {}
}
```

As far as the rest goes, you're probably asking the wrong person. I think anything but constructor based injection is non-sense. That opinion may be extreme, but I would point out that the authors of Spring (who AFAIK came up with the silly idea to allow for the rest) seem to advocate a similar position (scroll a bit down to "Constructor-based or setter-based DI?") in [this document](http://docs.spring.io/spring/docs/current/spring-framework-reference/html/beans.html#beans-setter-injection).

Cutting down verbosity is not a valid use case for setter based injection. Certainly not in a language with macros ;)

Optional dependencies would arguably be a valid use case for setter based injection, but I believe that you should not be having those either, because it's an indicator of the object has some fallback mechanism to deal with things that are outside its responsibility.

Look at it this way: a class designed with dependency inversion in mind but not for usage with a specific injector can always be used with it, because you can use Context.addGlobalMeta or some other macro magic to add the metadata externally. A class that is not usable without a specific framework will severely box you in.

**TLDR:** Use some macro (e.g. tink_lang) to cut down on verbosity. Don't use setter injection. Ever ;)

**Comment by jasononeil** [comment]
Good point Juraj.

I think I've been thinking similarly, and recently have mostly been using constructor injection, even when using minject, because it means your object is always in a valid state.  (You don't have a broken state where new() has been called, but dependencies haven't yet been injected.)

And dodrugs actually doesn't even bother looking for `@inject` metadata on the constructor, it assumes the constructor needs injection.  So it could be used to construct classes without any `@inject` metadata at all.

The next question is, do we:
1. Be opinionated and say "constructor injection only"
2. Keep supporting minject style injection points.

![why not both](https://cloud.githubusercontent.com/assets/315158/16822360/36c95dcc-499f-11e6-80e9-c03e0ed72d42.gif)

``` haxe
// Do something like tink constructor initialisation
@:autoBuild(...)interface Injectable {}

class Person implements Injectable {
    @inject("name") var name:String;
    @inject("age") var age:Int;
    @inject var favouriteNumbers:Array<Int>;
    var personality:Personality;

    @inject function injectPersonality(personalityApi:PersonalityApi) {
        personality = personalityApi.get();
    }

    @post function sayHello() {
        personality.introduceMyself(name);
    }
}

// generates a constructor:
class Person {
    @inject("name","age","","")
    public function new(name:String, age:Int, favouriteNumbers:Array<Int>, personalityApi:PersonalityApi,) {
        this.name = name;
        this.age = age;
        this.favouriteNumbers = favouriteNumbers
        this.injectPersonality( personalityApi );
        this.sayHello();
    }
}
```

The build macro will:
- [ ] Move all injections to the constructor
- [ ] Add `@inject` to the constructor, and remove it from the other points, so it'll still work with minject.

I mostly like the idea. 

Advantages:
- dodrugs.Injector will be able to work on normal classes without specific `@inject` metadata. So the class doesn't have to know the injector exists.
- It'll also work seamlessly with tink_lang initialisation.
- Will allow private injection when you use the Injectable interface.
- Objects using `Injectable` will be transformed in a way that is still compatible with minject.

Downsides:
- You'll need to add `implements Injectable` to classes that use the old style injection points.

**Comment by jasononeil** [comment]
I've ended up going with Juraj's suggestion and implementing constructor injection only.  Pull request welcome if someone wants to make a build macro that turns minject style injection points into the constructor injection style.  :)
